mitiq/cdr/cdr_execution.py

```diff
@@ -0,0 +1,179 @@
+                                    ``method_select='gaussian'``.
+            - sigma_replace (float): Width of the Gaussian distribution used
+                                     for ``method_replace='gaussian'``.
+            - random_state (int): Seed for sampling.
+
+    Returns: The tuple (raw_expectations, mitigated_expectations)
+             corresponding to the many raw expectation values (at different
+             noise levels) and the associated mitigated values.
+
+    .. [Czarnik2020] : Piotr Czarnik, Andrew Arramsmith, Patrick Coles,
+        Lukasz Cincio, "Error mitigation with Clifford quantum circuit
+        data," (https://arxiv.org/abs/2005.10189).
+    .. [Lowe2020] : Angus Lowe, Max Hunter Gordon, Piotr Czarnik,
+        Andrew Arramsmith, Patrick Coles, Lukasz Cincio,
+        "Unified approach to data-driven error mitigation,"
+        (https://arxiv.org/abs/2011.01157).
+    """
+    # Handle keyword arguments for generating training circuits.
+    method_select = kwargs.get("method_select", "uniform")
+    method_replace = kwargs.get("method_replace", "closest")
+    random_state = kwargs.get("random_state", None)
+    kwargs_for_training_set_generation = {
+        "sigma_select": kwargs.get("sigma_select"),
+        "sigma_replace": kwargs.get("sigma_replace"),
+    }
+
+    # Generate training circuits.
+    training_circuits = generate_training_circuits(
+        circuit,
+        num_training_circuits,
+        fraction_non_clifford,
+        method_select,
+        method_replace,
+        random_state,
+        kwargs=kwargs_for_training_set_generation,
+    )
+
+    # [Optionally] Scale noise in circuits.
+    all_circuits = [
+        [scale_noise(c, s) for s in scale_factors]
+        for c in [circuit] + training_circuits
+    ]
+
+    # Execute all circuits. TODO: Allow for batching.
+    executor_data = np.array(
+        [[executor(circ) for circ in circuits] for circuits in all_circuits]
+    )
+    simulator_data = np.array([simulator(circ) for circ in all_circuits[0]])
+
+    # Do the regression.
+    results_dict_training_circuits = [simulator_data, executor_data]
+    results_dict_circuit_of_interest = executor_data[:, 0]
+
+    # Now the regression:
+    mitigated_observables = []
+    raw_observables = []
+    for obs in observables:
+        circuit_data = construct_circuit_data_floats(
+            results_dict_circuit_of_interest, obs
+        )
+        train_data = construct_training_data_floats(
+            results_dict_training_circuits, obs
+        )
+        # going to add general regression section here:
+        fitted_params, _ = curve_fit(
+            lambda x, *params: ansatz(x, params),
+            train_data[0].T,
+            train_data[1],
+            p0=np.zeros(num_parameters),
+        )
+        mitigated_observables.append(ansatz(circuit_data, fitted_params))
+        raw_observables.append(circuit_data)
+
+    return raw_observables, mitigated_observables
```

Comment:
If you agree with the previous comment these should be switched (here and in tests):
```suggestion
    return mitigated_observables, return raw_observables
```